hls-plugin-api/src/Ide/Types.hs

```diff
@@ -877,6 +886,57 @@ type CommandFunction ideState a
 
 -- ---------------------------------------------------------------------
 
+type ResolveFunction ideState a (m :: Method ClientToServer Request) =
+  ideState
+  -> PluginId
+  -> MessageParams m
+  -> Uri
+  -> a
+  -> LspM Config (Either ResponseError (MessageResult m))
+
+-- | Make a handler for plugins with no extra data
+mkResolveHandler
+  :: forall ideState a m. (FromJSON a,  PluginRequestMethod m, L.HasData_ (MessageParams m) (Maybe Value))
+  =>  SClientMethod m
+  -> (ideState
+  ->PluginId
+  -> MessageParams m
+  -> Uri
+  -> a
+  -> LspM Config (Either ResponseError (MessageResult m)))
+  -> PluginHandlers ideState
+mkResolveHandler m f = mkPluginHandler m f'
+  where f' ideState plId params = do
+          case fromJSON <$> (params ^. L.data_) of
+            (Just (Success (PluginResolveData owner uri value) )) -> do
+              if owner == plId
+              then
+                case fromJSON value of
+                  Success decodedValue ->
+                    let newParams = params & L.data_ ?~ value
+                    in f ideState plId newParams uri decodedValue
+                  Error err ->
+                    pure $ Left $ ResponseError (InR ErrorCodes_ParseError) (parseError value err) Nothing
+              else pure $ Left $ ResponseError (InR ErrorCodes_InvalidRequest) invalidRequest Nothing
+            _ -> pure $ Left $ ResponseError (InR ErrorCodes_InvalidRequest) invalidRequest Nothing
```

Comment:
use the aeson error!

Reply:
Also I don't think this should be using the "invalidRequest" message?

Reply:
Yeah, maybe InternalError would work better here